source/simulator.cpp

```diff
@@ -0,0 +1,126 @@
+#include "simulator.hpp"
+
+#include <memory>
+#include <set>
+
+#include "device.hpp"
+#include "event.hpp"
+#include "logger.hpp"
+#include "receiver.hpp"
+#include "sender.hpp"
+#include "switch.hpp"
+
+namespace sim {
+
+Simulator::Simulator() : m_scheduler(Scheduler::get_instance()) {}
+
+std::shared_ptr<IRoutingDevice> Simulator::add_device(std::string a_name,
+                                                      DeviceType a_type) {
+    if (m_graph.find(a_name) != m_graph.end()) {
+        LOG_WARN("add_device failed: device with name {} already exists.",
+                     a_name);
+        return nullptr;
+    }
+    switch (a_type) {
+        case DeviceType::SENDER:
+            m_graph[a_name] = std::make_shared<Sender>();
+            break;
+        case DeviceType::SWITCH:
+            m_graph[a_name] = std::make_shared<Switch>();
+            break;
+        case DeviceType::RECEIVER:
+            m_graph[a_name] = std::make_shared<Receiver>();
+            break;
+    }
+    return m_graph[a_name];
+}
+
+void Simulator::add_flow(std::shared_ptr<IRoutingDevice> a_from,
+                         std::shared_ptr<IRoutingDevice> a_to) {
+    ISender* sender = dynamic_cast<ISender*>(a_from.get());
+    IReceiver* receiver = dynamic_cast<IReceiver*>(a_to.get());
+    if (sender == nullptr) {
+        LOG_WARN("add_flow failed: sender is nullptr");
+        return;
+    }
+    
+    if (receiver== nullptr) {
+        LOG_WARN("add_flow failed: receiver is nullptr");
+        return;
+    }
+    m_flows.emplace_back(std::make_shared<Flow>(sender, receiver, 0.f));
+}
+
+void Simulator::add_link(std::shared_ptr<IRoutingDevice> a_from,
+                         std::shared_ptr<IRoutingDevice> a_to,
+                         std::uint32_t a_speed_mbps, std::uint32_t a_delay) {
```

Comment:
Use Time instead of std::uint32_t
```suggestion
                         std::uint32_t a_speed_mbps, Time a_delay) {
```